src/main/java/com/xpn/xwiki/plugin/lucene/searcherProvider/SearcherProvider.java

```diff
@@ -38,82 +41,72 @@
     LOGGER.debug("create searcherProvider: [" + System.identityHashCode(this) + "].");
   }
 
-  HashMap<Thread, Set<SearchResults>> internal_getConnectedSearchResults() {
+  ConcurrentMap<Thread, Set<SearchResults>> internal_getConnectedSearchResults() {
     return connectedSearchResultsMap;
   }
 
   Set<Thread> internal_getConnectedThreads() {
     return connectedThreads;
   }
 
-  /**
-   * connect and markToClose need to be externally synchronized
-   */
   public void connect() {
     if (!checkConnected()) {
-      if (this.markToClose) {
-        throw new IllegalStateException(
-            "you may not connect to a SearchProvider" + " marked to close.");
-      }
-      LOGGER.debug("connect searcherProvider [" + System.identityHashCode(this) + "] to ["
-          + Thread.currentThread() + "].");
+      checkState(!markToClose, "you may not connect to a SearchProvider marked to close.");
+      LOGGER.debug("connect searcherProvider [{}] to [{}].", System.identityHashCode(this),
+          Thread.currentThread());
       connectedThreads.add(Thread.currentThread());
     }
   }
 
   public boolean isClosed() {
-    return (this.searchers == null);
+    return this.isClosed;
   }
 
   public Searcher[] getSearchers() {
-    if (!checkConnected()) {
-      throw new IllegalStateException("you must connect to the searcher provider before"
-          + " you can get any searchers");
-    }
-    return this.searchers;
+    checkState(!isClosed(), "Getting serachers failed: provider is closed.");
+    checkState(checkConnected(), "you must connect to the searcher provider before you can get"
+        + " any searchers");
+    return backedSearchers;
   }
 
   private boolean checkConnected() {
-    return this.connectedThreads.contains(Thread.currentThread());
+    return connectedThreads.contains(Thread.currentThread());
   }
 
   public void disconnect() throws IOException {
     if (connectedThreads.remove(Thread.currentThread())) {
-      LOGGER.debug(
-          "disconnect searcherProvider [" + System.identityHashCode(this) + "] to ["
-              + Thread.currentThread() + "], markedToClose [" + isMarkedToClose() + "].");
+      LOGGER.debug("disconnect searcherProvider [{}] to [{}], markedToClose [{}].",
+          System.identityHashCode(this), Thread.currentThread(), isMarkedToClose());
       closeIfIdle();
     }
   }
 
   public boolean isMarkedToClose() {
-    return this.markToClose;
+    return markToClose;
   }
 
-  /**
-   * connect and markToClose need to be externally synchronized
-   */
   public void markToClose() throws IOException {
-    if (!this.markToClose) {
-      LOGGER
-          .debug("markToClose searcherProvider [" + System.identityHashCode(this) + "].");
-      this.markToClose = true;
+    if (!isMarkedToClose()) {
+      LOGGER.debug("markToClose searcherProvider [{}].", System.identityHashCode(this));
+      markToClose = true;
```

Comment:
`markToClose = true` should be in a synchronized block, see comment above for `connect()`